lib/src/json_schema/custom_vocabularies.dart

```diff
@@ -0,0 +1,114 @@
+//     of this software and associated documentation files (the "Software"), to deal
+//     in the Software without restriction, including without limitation the rights
+//     to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//     copies of the Software, and to permit persons to whom the Software is
+//     furnished to do so, subject to the following conditions:
+//
+//     The above copyright notice and this permission notice shall be included in
+//     all copies or substantial portions of the Software.
+//
+//     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//     AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//     THE SOFTWARE.
+
+import 'package:json_schema/json_schema.dart';
+
+/// Use to register a custom vocabulary with the [JsonSchema] compiler.
+///
+class CustomVocabulary {
+  CustomVocabulary(this._vocab, this._keywordImplementations);
+
+  Uri _vocab;
+  Map<String, CustomKeywordImplementation> _keywordImplementations;
+
+  /// Name of the vocabulary.
+  Uri get vocab => _vocab;
+
+  /// A map of the keywords and implementation for the keywords.
+  Map<String, CustomKeywordImplementation> get keywordImplementations => _keywordImplementations;
+}
+
+/// A class to contain the set of functions for setting and validating keywords in a custom vocabulary.
+///
+/// The two functions provided are used to process an attribute in a schema and then validate data.
+///
+/// The setter function takes the current JsonSchema node being processed and the data from the json.
+/// The given function should validate and transform the data however is needed for the corresponding validation
+/// function. If the data is bad a [FormatException] with a clear message should be thrown.
+///
+/// The validation function takes the output from the property setter and data from a JSON payload to be validated.
+/// A [CustomValidationResult] should be returned to indicate the outcome of the validation.
+class CustomKeywordImplementation {
+  CustomKeywordImplementation(this._propertySetter, this._validator);
+
+  Object Function(JsonSchema schema, Object value) _propertySetter;
+  CustomValidationResult Function(Object schemaProperty, Object instanceData) _validator;
+
+  /// Function used to set a property from the a schema.
+  Object Function(JsonSchema schema, Object value) get propertySetter => this._propertySetter;
+
+  /// Function used to validate a json value.
+  CustomValidationResult Function(Object schemaProperty, Object instanceData) get validator => this._validator;
+}
+
+enum _ValidationState { valid, warning, error }
```

Comment:
Should we document that `warning` means "valid, but FYI things smell"?

Reply:
I added some documentation around the class.